src/platforms/esp32/main/spi_driver.c

```diff
@@ -68,91 +127,113 @@ void spi_driver_init(GlobalContext *global)
     buscfg.quadwp_io_num = -1;
     buscfg.quadhd_io_num = -1;
 
-    term device_config = interop_proplist_get_value(opts, DEVICE_CONFIG_ATOM);
-    term clock_speed_hz_term = interop_proplist_get_value(device_config, SPI_CLOCK_HZ_ATOM);
-    term mode_term = interop_proplist_get_value(device_config, SPI_MODE_ATOM);
-    term spics_io_num_term = interop_proplist_get_value(device_config, SPI_CS_IO_NUM_ATOM);
-    term address_bits_term = interop_proplist_get_value(device_config, ADDRESS_LEN_BITS_ATOM);
+    debug_buscfg(&buscfg);
 
-    spi_device_interface_config_t devcfg;
-    memset(&devcfg, 0, sizeof(spi_device_interface_config_t));
-    devcfg.clock_speed_hz = term_to_int32(clock_speed_hz_term);
-    devcfg.mode = term_to_int32(mode_term);
-    devcfg.spics_io_num = term_to_int32(spics_io_num_term);
-    devcfg.queue_size = 4;
-    devcfg.address_bits = term_to_int32(address_bits_term);
-
-    int ret = spi_bus_initialize(HSPI_HOST, &buscfg, 1);
-
-    if (ret == ESP_OK) {
-        TRACE("initialized SPI\n");
+    esp_err_t err = spi_bus_initialize(host_device, &buscfg, 1);
+    if (err == ESP_OK) {
+        ESP_LOGI(TAG, "SPI Bus initialized.");
     } else {
-        TRACE("spi_bus_initialize return code: %i\n", ret);
+        ESP_LOGE(TAG, "SPI Bus initialization failed with error=%i", err);
+        context_destroy(ctx);
+        return NULL;
     }
 
-    ret = spi_bus_add_device(HSPI_HOST, &devcfg, &spi_data->handle);
-
-    if (ret == ESP_OK) {
-        TRACE("initialized SPI device\n");
-    } else {
-        TRACE("spi_bus_add_device return code: %i\n", ret);
+    struct SPIData *spi_data = calloc(1, sizeof(struct SPIData));
+    list_init(&spi_data->devices);
+    // TODO handle out of memory errors
+    spi_data->host_device = host_device;
+
+    term device_map = term_get_map_assoc(ctx, opts, DEVICE_CONFIG_ATOM);
+    term device_names = term_get_map_keys(device_map);
+
+    int n = term_get_map_size(device_map);
+    for (int i = 0; i < n; ++i) {
+        term device_name = term_get_tuple_element(device_names, i);
+        term device_config = term_get_map_assoc(ctx, device_map, device_name);
+
+        term clock_speed_hz_term = term_get_map_assoc(ctx, device_config, SPI_CLOCK_HZ_ATOM);
+        term mode_term = term_get_map_assoc(ctx, device_config, SPI_MODE_ATOM);
+        term spics_io_num_term = term_get_map_assoc(ctx, device_config, SPI_CS_IO_NUM_ATOM);
+        term address_bits_term = term_get_map_assoc(ctx, device_config, ADDRESS_LEN_BITS_ATOM);
+
+        spi_device_interface_config_t devcfg = { 0 };
+        devcfg.clock_speed_hz = term_to_int32(clock_speed_hz_term);
+        devcfg.mode = term_to_int32(mode_term);
+        devcfg.spics_io_num = term_to_int32(spics_io_num_term);
+        devcfg.queue_size = 4;
+        devcfg.address_bits = term_to_int32(address_bits_term);
+
+        spi_device_handle_t handle;
+        err = spi_bus_add_device(host_device, &devcfg, &handle);
+        if (err == ESP_OK) {
```

Comment:
let's rewrite this as an early return when `err != ESP_OK`, I think it is slightly more readable.